eth/watchers/unbondingwatcher.go

```diff
@@ -0,0 +1,162 @@
+package watchers
+
+import (
+	"fmt"
+	"math/big"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+)
+
+type unbondingLockStore interface {
+	InsertUnbondingLock(id *big.Int, delegator ethcommon.Address, amount, withdrawRound *big.Int) error
+	DeleteUnbondingLock(id *big.Int, delegator ethcommon.Address) error
+	UseUnbondingLock(id *big.Int, delegator ethcommon.Address, usedBlock *big.Int) error
+}
+
+// UnbondingWatcher watches for on-chain events to update the state of an unbonding lock store
+type UnbondingWatcher struct {
+	addr  ethcommon.Address // Watching for on-chain events pertaining to this address
+	bw    blockWatcher
+	store unbondingLockStore
+	dec   *EventDecoder
+	sub   ethereum.Subscription
+
+	quit chan struct{}
+}
+
+// NewUnbondingWatcher creates an UnbondingWatcher instance
+func NewUnbondingWatcher(addr ethcommon.Address, bondingManagerAddr ethcommon.Address, bw blockWatcher, store unbondingLockStore) (*UnbondingWatcher, error) {
+	dec, err := NewEventDecoder(bondingManagerAddr, contracts.BondingManagerABI)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UnbondingWatcher{
+		addr:  addr,
+		bw:    bw,
+		store: store,
+		dec:   dec,
+		quit:  make(chan struct{}),
+	}, nil
+}
+
+// Start kicks off a loop that handles events from a block subscription
+func (w *UnbondingWatcher) Start() {
+	blockEvents := make(chan []*blockwatch.Event, 10)
+	w.sub = w.bw.Subscribe(blockEvents)
+
+	for {
+		select {
+		case <-w.quit:
+			w.sub.Unsubscribe()
+			close(blockEvents)
```

Comment:
I don't think we should be closing the sink channel from the receiving side. I think unsubscribe will close the producer channel. This could lead to unwanted panicking by closing an already closed channel. 

```
// NewSubscription runs a producer function as a subscription in a new goroutine. The
// channel given to the producer is closed when Unsubscribe is called. If fn returns an
// error, it is sent on the subscription's error channel.
```

Reply:
The subscription here is actually the result of calling [(*Feed).Subscribe()](https://github.com/ethereum/go-ethereum/blob/master/event/feed.go) which accepts an `interface{}` channel which allows our sink channel to be of any event type. Under the hood, the subscription returned is implemented by `feedSub` and we can see [here](https://github.com/ethereum/go-ethereum/blob/master/event/feed.go#L199) that `Unsubscribe()` *does not* close the sink channel. Instead, `Unsubscribe()` closes the subscription's *error channel* and tells `*Feed` to [remove](https://github.com/ethereum/go-ethereum/blob/master/event/feed.go#L104) the subscription from its known set of subscriptions. When removing a subscription, `*Feed` does not touch the original sink channel - it just removes the subscription so that whenever the feed receives a new event, it will not send to the sink channel for that subscription. 

Re: the `NewSubscription` comment - note that the channel accepted by the producer function is a `struct{}` type. This channel is not meant to be used as the sink channel (which we know need to be able to accept any event type). The channel accepted by the producer is actually used to signal to the producer function that `Unsubscribe()` was called and is not used to carry values to be emitted for the subscription.

Reply:
I see what you are saying. But wouldn't closing the sink channel from the receiver side lead to potential unwanted behaviour? 

Isn't it an unwritten rule to only close channels from the sender side? Not sure anymore where I got that from. 

Altough I guess if we remove the subscription the sink channel the sender will not be sending values into the channel anymore so we can safely close it.

Reply:
> Isn't it an unwritten rule to only close channels from the sender side?

After a second thought, I do agree with you. It is generally a good convention to only have senders close channels to avoid the sending on closed channels.

My original desire to make sure to close the channel after it is done being used was misplaced - there actually isn't any need to close a channel because if it is not being used GC will clean it up. The primary reason for closing a channel is to signal to a consumer that no more values will be emitted. In this case, since the consumer is the one that decides it doesn't care about values anymore, it also doesn't need a signal from a closed channel.

> Although I guess if we remove the subscription the sink channel the sender will not be sending values into the channel anymore so we can safely close it.

While this is true, I think you make a good point about following the convention of only having senders close channels. Thanks for bringing it up.

Fixed in b680940